packages/block-editor/src/store/reducer.js

```diff
@@ -2287,15 +2299,41 @@ function getDerivedBlockEditingModesForTree(
 
 			// If zoomed out, all blocks that aren't sections or the section root are
 			// disabled.
-			// If the tree root is not in a section, set its editing mode to disabled.
-			if (
-				isZoomedOut ||
-				! findParentInClientIdsList( state, clientId, sectionClientIds )
-			) {
+			if ( isZoomedOut ) {
 				derivedBlockEditingModes.set( clientId, 'disabled' );
 				return;
 			}
 
+			const isInSection = !! findParentInClientIdsList(
+				state,
+				clientId,
+				sectionClientIds
+			);
+			if ( ! isInSection ) {
+				if ( clientId === '' ) {
+					derivedBlockEditingModes.set( clientId, 'disabled' );
+					return;
+				}
+
+				// Allow selection of template parts outside of sections.
+				if ( blockName === 'core/template-part' ) {
```

Comment:
I'm curious - are we "allowed" to reference core blocks like this in the block editor package?

If not how can we pass in references to these blocks and express the custom logic here that relates to "template parts"?

I was thinking of a private block editor setting, but then I'm not sure how we'd handle the custom logic that seems to relate specifically to template parts.

I'm also thinking about how we could extend this to the Navigation block in the future.

Reply:
I think not ideal to have loads of this, but it's ok for testing new concepts. I don't think it's harmful to any non-WP consumers of the block editor package as these bits of code will have no effect. It's just a bit of a code quality issue to have lots of hard to follow conditions, and it's better in the long run to try unifying some concepts behind an API or similar, but sometimes the right one takes a while to emerge.

Hopefully someday in the future template parts will be deprecated and there will only be synced patterns, and they'll work the same way.

Reply:
We do have a fair bit of this already in the block editor package 😅 

Could we use a similar logic to get/set `groupingBlockName` for the template part block? Is it likely that third parties will provide other types of template part-like blocks that we might want this code to work with?